### PR TITLE
Make EM_ASM an error in strict C mode

### LINKED ISSUE #8860: EM_ASM is broken without extra parameters
The following code used to work:
```c
	EM_ASM({
		var sdl2 = Module['SDL2'];
		if(sdl2 !== undefined) {
			var audioctx = sdl2.audioContext;
			if(audioctx !== undefined) {
				audioctx.resume();
			}
		}
	});
```

After updating to latest emscripten I get the following: https://gist.githubusercontent.com/Akaricchi/e970633fdfd8816338574c7e096f5428/raw/28b0ec9104ab9c56f6309ee6a1de5255a121c5cb/gistfile1.txt

Appending a dummy unused argument appears to work, although it still generates a -Wgnu-zero-variadic-macro-arguments warning, annoyingly enough — maybe you should consider suppressing it with `__extension__` so that users wouldn't have to disable the warning or do that in their own code. `#define EM_ASM(...) (__extension__ (.. the rest ..))` should do the trick.

### PR DESCRIPTION
**Pull request by quantum5**
Fixes #8860.

### CODE REVIEW

File: system/include/emscripten/em_asm.h [hunk 1]
```diff
@@ -7,6 +7,9 @@
 
 #pragma once
 
+// EM_ASM does not work strict C mode.
+#if defined(__cplusplus) || !defined(__STRICT_ANSI__)
```
Comment (jgravelle-google) [inline system/include/emscripten/em_asm.h]: Can we invert the conditional here?
`#if !defined(__cplusplus) && defined(__STRICT_ANSI__)`

That moves the error definitions much closer to the comment...

File: system/include/emscripten/em_asm.h [hunk 2]
```diff
@@ -214,3 +216,20 @@ void emscripten_asm_const_async_on_main_thread(const char* code, ...);
 #define EM_ASM_ARGS(code, ...) emscripten_asm_const_int(#code _EM_ASM_PREP_ARGS(__VA_ARGS__))
 #define EM_ASM_INT_V(code) EM_ASM_INT(#code)
 #define EM_ASM_DOUBLE_V(code) EM_ASM_DOUBLE(#code)
+
+#else
```
Comment (jgravelle-google) [inline system/include/emscripten/em_asm.h]: ... while not needing to remind the reader, 200 lines later, what the "else" condition here means.

File: system/include/emscripten/em_asm.h [hunk 3]
```diff
@@ -214,3 +216,20 @@ void emscripten_asm_const_async_on_main_thread(const char* code, ...);
...
+#define MAIN_THREAD_EM_ASM(...) EM_ASM_ERROR
+#define MAIN_THREAD_EM_ASM_INT(...) EM_ASM_ERROR
+#define MAIN_THREAD_EM_ASM_DOUBLE(...) EM_ASM_ERROR
+#define MAIN_THREAD_ASYNC_EM_ASM(...) EM_ASM_ERROR
+#define EM_ASM_(...) EM_ASM_ERROR
+#define EM_ASM_ARGS(...) EM_ASM_ERROR
+#define EM_ASM_INT_V(...) EM_ASM_ERROR
+#define EM_ASM_DOUBLE_V(...) EM_ASM_ERROR
+
+#endif
```
Comment (jgravelle-google) [inline system/include/emscripten/em_asm.h]: As a general rule I find it useful to comment what the elses/endifs refer to, so it's easier to trace which level of nesting we're on, hundreds of lines later.

**Comment by kripken** [comment]
This broke on windows, https://logs.chromium.org/logs/emscripten-releases/buildbucket/cr-buildbucket.appspot.com/8906912674451776064/+/steps/Emscripten_testsuite__upstream__other_/0/stdout

(why does the test use pty, seems like it can use run_process like other tests?)

File: tests/test_other.py [hunk 4]
```diff
@@ -9552,3 +9552,14 @@ def test_assertions_on_incoming_module_api_changes(self):
...
     self.assertContained('Module.read option was removed', run_js('a.out.js', assert_returncode=None, stderr=PIPE))
+
+  def test_em_asm_strict_c(self):
+    create_test_file('src.c', '''
+      #include <emscripten/em_asm.h>
+      int main() {
+        EM_ASM({ console.log('Hello, world!'); });
+      }
+    ''')
+    returncode, output = self.run_on_pty([PYTHON, EMCC, '-std=c11', 'src.c'])
```
Comment (yeputons) [inline tests/test_other.py]: Some other tests use `self.expect_fail` (e.g. `test_EM_ASM_i64`)

**Pull request merged**